### PR TITLE
cgroups: Add support to dump fake attach events

### PR DESCRIPTION
**Pull request by joelagnel**
With this its possible for users to dump the initial cgroup state into
the traces with something like the following:

cgroup = te.target.cgroups.controllers['cpuset'].cgroup('/foreground')
cgroup.trace_cgroup_tasks()

Signed-off-by: Joel Fernandes <joelaf@google.com>

### CODE REVIEW

**Comment by joelagnel** [comment]
CC @derkling @bjackman

**Comment by bjackman** [comment]
The shutil and `trace_cgroup_tasks` look good to me. 

I don't understand why `create_existing_cgroups` is added? I might be being slow here...

Also could you add a docstring explaining what `trace_existing_cgroups` does?

**Comment by joelagnel** [comment]
I think it would work without this too. But it made sense to me to pre create the cgroup. Now that you mention this I am thinking I'll drop it and just build it later.

**Comment by joelagnel** [comment]
@derkling please merge this?

File: devlib/bin/scripts/shutils.in [hunk 1]
```diff
@@ -177,6 +177,15 @@ cgroups_tasks_in() {
 	exit 0
 }
 
+cgroup_trace_attach_task() {
+        DST_ROOT=${1}
+        DST_PATH=${2}
+        TASKS_FILE=${3}
+
+        cat $TASKS_FILE | while read PID; do
+            echo "cgroup_attach_task_devlib: dst_root=$DST_ROOT dst_path=$DST_PATH pid=$PID" > /sys/kernel/debug/tracing/trace_marker
```
Comment (derkling) [inline devlib/bin/scripts/shutils.in]: Maybe this format is less aggressive on the 80 cols limit?
```python
EVENT= "cgroup_attach_task_devlib: dst_root=$DST_ROOT dst_path=$DST_PATH pid=$PID"
cat $TASKS_FILE | while read PID; do
    echo $EVENT >/sys/kernel/debug/tracing/trace_marker
done
```

Moreover, this is not the same format of the original (i.e. kernel generated) event... but I guess we cannot do otherwise since we don't know the source groups.

File: devlib/module/cgroups.py [hunk 2]
```diff
@@ -287,6 +287,11 @@ def get_tasks(self):
         logging.debug('Tasks: %s', task_ids)
         return map(int, task_ids)
 
+    # Used to insert fake cgroup attach events to know existing cgroup assignments
+    def trace_cgroup_tasks(self):
```
Comment (derkling) [inline devlib/module/cgroups.py]: We are already into the CGroup class, thous I would suggest to rename this to be just `CGroup::trace_tasks`

Comment (derkling) [inline devlib/module/cgroups.py]: Moreover, I think we should also add a method to the `CgroupsModule` class, which allows to trace the cgroups for all the CGroups of all the currently mounted controllers.
I think this will be a quite common use case and we don't want to have this code in the clients of the CGroups API. Possibly, this new call should also allow to specify a "mask" of cgroups names to report. Something like:

```python
class CgroupModule(Module):

    def trace_cgroups(self, controllers=None, cgroups=None):
        if controllers is None:
            controllers = [ss.name for ss in self.controllers]
        for controller in self.controllers:
            if controller not in controllers:
                continue
            controller.trace_cgroups(cgroups)
```

Which ultimately requires an additional method in the `Contoller` class doing almost the same but filtering on cgroups names to call your method above just on listed cgroups (or all if None).

If you don't have time, I can provide such a patch in the next few days...

File: devlib/module/cgroups.py [hunk 3]
```diff
@@ -287,6 +287,11 @@ def get_tasks(self):
...
+        exec_cmd = "cgroup_trace_attach_task {} {} {}".format(self.controller.hid, self.directory, self.tasks_file)
```
Comment (derkling) [inline devlib/module/cgroups.py]: you can save some columns by splitting before `.format`

File: devlib/bin/scripts/shutils.in (hunk 1 above)
Comment (joelagnel) [inline devlib/bin/scripts/shutils.in]: Yes I believe, only the extra fields are dropped but the existing fields have the same format, so that's not an issue

File: devlib/module/cgroups.py (hunk 2 above)
Comment (joelagnel) [inline devlib/module/cgroups.py]: Honestly I think this can be a wrapper on top of my patch. I do believe my patch is clean enough but all that's missing is a wrapper to take controller and group parameters. IMO the core dumping into trace logic should be as low as possible (infact right fully in the `CGroup` class since we're dumping tasks per `CGroup`). Also I tried implementing it the other way - make the `CGroup controller` or `CgroupModule` do the dumping into trace and it complicates things because `self.tasks_file` is available in `CGroup` class and can just be directly accessed and passed along to the `shutil`. So I still like my approach to this is clean.

If you still want to do this differently, I am Ok with that. But I would like it to be as simple as my approach. Maybe you can just add a wrapper to this patch in `CgroupModule` if your concern is the interface?

File: devlib/module/cgroups.py (hunk 3 above)
Comment (joelagnel) [inline devlib/module/cgroups.py]: Ok.

File: devlib/bin/scripts/shutils.in (hunk 1 above)
Comment (joelagnel) [inline devlib/bin/scripts/shutils.in]: Agreed with comment on 80 line limit and will fix, thanks

File: devlib/module/cgroups.py (hunk 2 above)
Comment (derkling) [inline devlib/module/cgroups.py]: My concern is that, since we are adding a new API, is should be properly implemented to be fully usable... while this one is a "partial implementation".

I guess your code which uses this API is already doing most of the things I'm asking for here, at least looping on all cgroups of a controller. Since this is a quite common use case, I would like we merge the API once we provide these minimal set of functionalities covered.

To summarise, what you did is simple, clean and ok... but a bit limited in scope to require additional client code. Let's try to factor in this client code in the new API.

If you do not have time to provide such an extension it's ok for me to add some patches on top of this PR before merging it.

Comment (joelagnel) [inline devlib/module/cgroups.py]: Yes, I think if you can add some PRs on top of this with your idea, that will be the quickest. I think the high level implementation can be simple wrapper to this lower level functionality. Thanks @derkling

**Comment by derkling** [comment]
Hi @joelagnel, you can find here:
   derkling:review-joel-for-cgroup-attach
a series with few patches I've added on top of the one of this PR.

Modifications:
- fixup! cgroups: Add support to dump fake attach events
  This should be squased into your own... it brings few updates on your original patch
- cgroups: add high level API for controller/cgroups tracing
  Introduces the high-level API I've suggested in one of my previous comments
- Revert "ssh: Combine exit code and main command"
  This reverts an optimisation by @bjackman , I would not add it to your PR, since it's not affecting 
  Android targets... but I needed it to test on my Juno/Linux target.

Please give a try to these new bits... especially the high-level API.

**Comment by setrofim** [comment]
@derkling @joelagnel @bjackman what is the state of this PR? Is this still needed? Are there more changes coming?

**Comment by joelagnel** [comment]
These are needed. I will squash Patrick's fixup patch into mine and submit
it, sorry about the delay. This PR was already in our tree and is being
used for some time without the need for the high level API.

**Comment by douglas-raillard-arm** [comment]
Closing as this has been sitting there for 7.5 years and we now have a new cgroups implementation (cgroups2 devlib module), with the existing one being soft-deprecated.
